### PR TITLE
[Caffe2] Added Top_k operator

### PR DESCRIPTION
**Pull request by ashishfarmer**

### CODE REVIEW

**Comment by ashishfarmer** [comment]
@iotamudelta, I made some changes in THCAsmUtils.cuh. Please have a look.
Those functions are LaneMask functions and must return a 64 bit value for HIP/HCC path

**Comment by ashishfarmer** [comment]
Another instance of `test_Conv2d_groups_thnn (__main__.TestNN)` failure

File: caffe2/operators/top_k_heap_selection.cuh [hunk 1]
```diff
@@ -4,7 +4,11 @@
 #include "caffe2/utils/GpuBitonicSort.cuh"
 #include "caffe2/utils/GpuDefs.cuh"
 #include "caffe2/utils/math.h"
+#if defined(__HIP_PLATFORM_HCC__)
+#include <hip/hip_runtime.h>
```
Comment (iotamudelta) [inline caffe2/operators/top_k_heap_selection.cuh]: Is this necessary? Automatic hipification should work here?

File: caffe2/operators/top_k_radix_selection.cuh [hunk 2]
```diff
@@ -4,7 +4,11 @@
 #include "caffe2/utils/GpuDefs.cuh"
 #include "caffe2/utils/GpuScanUtils.cuh"
 #include "caffe2/utils/math.h"
+#if defined(__HIP_PLATFORM_HCC__)
+#include <hip/hip_runtime.h>
```
Comment (iotamudelta) [inline caffe2/operators/top_k_radix_selection.cuh]: Same here.

File: tools/amd_build/build_caffe2_amd.py [hunk 3]
```diff
@@ -26,7 +26,6 @@
 ignores = [
     "caffe2/operators/depthwise_3x3_conv_op.cu",
     "caffe2/operators/depthwise_3x3_conv_op_cudnn.cu",
-    "caffe2/operators/top_k.cu",
     "caffe2/operators/top_k_radix_selection.cuh",
```
Comment (iotamudelta) [inline tools/amd_build/build_caffe2_amd.py]: Purely out of interest: aren't these close enough to TopK to just get enabled as well?

**Comment by ashishfarmer** [comment]
Very valid points @iotamudelta . All the comments have one solution - to keep the .cuh files under pyhipify. Let me check and refactor this based on how we treat other cuh files in the Caff2 project

**Pull request merged**